lib/fastest-node-client.ts

```diff
@@ -0,0 +1,78 @@
+import {Chain, ChainClient, ChainOptions, defaultChainOptions, RandomnessBeacon} from './index'
+import HttpCachingChain, {HttpChain} from './http-caching-chain'
+import {createSpeedTest, SpeedTest} from './speedtest'
+import HttpChainClient from './http-chain-client'
+
+const defaultSpeedTestInterval = 1000 * 60
+
+type SpeedTestEntry = {
+    test: SpeedTest
+    url: string
+}
+
+// takes an array of drand nodes and periodically speed tests them to work out which is the fastest
+// it then uses the fastest client to make calls using an underlying HTTP client
+// use `.start()` to enable optimisation of the underlying base URLs
+// don't forget to `.stop()` it after you're finished
+class FastestNodeClient implements ChainClient {
+
+    speedTests: Array<SpeedTestEntry> = []
+
+    constructor(
+        public baseUrls: Array<string>,
+        public options: ChainOptions = defaultChainOptions,
+        private speedTestIntervalMs = defaultSpeedTestInterval
+    ) {
+        if (baseUrls.length === 0) {
+            throw Error('Can\'t optimise an empty `baseUrls` array!')
+        }
```

Comment:
It's also useless if the array is `=== 1`, no?

Reply:
heh good point! Perhaps `=== 1` should be warning though - I can imagine a situation where a single implementation pulls its lists of URLs from elsewhere and doesn't want to hotswap impls

Reply:
Have added a warning for this